### PR TITLE
Spider multiple markets

### PR DESCRIPTION
**Pull request by scott-silver**
Updating Spider (and the things that depend on Spider) to support multiple deployments on the same chain.

Making this update means that we'll frequently be adding an additional `--market` argument to many of the commands that previously just took a `--network` argument.

```
# before
yarn spider --network kovan

# after
yarn spider --network kovan --market 01-USDC
```

Our scripts (like `deploy-comet` and `verify-comet`) will require market to be passed in via an environment variable (`MARKET="01-USDC"`):

```
MARKET="01-USDC" yarn hardhat run scripts/deploy-comet --network kovan
```

---

Going to make two changes discussed earlier with 👨‍🍳  and 🍯:
- [x] rename "market" to "deployment"
- [x] delete the number prefix ("01-USDC")

### CODE REVIEW

File: .github/workflows/run-spider.yaml [hunk 1]
```diff
@@ -1,33 +0,0 @@
-name: 'Run Spider'
```
Comment (scott-silver) [inline .github/workflows/run-spider.yaml:1]: deleted, per your question @jflatow:

https://discord.com/channels/402910780124561410/796452409773129778/1002590359756480601

File: README.md [hunk 2]
```diff
@@ -228,7 +228,7 @@ Uniswap for a profit.
 To run the bot, you'll need the address of a deployed version of the Liquidator
 contract (or you can deploy a new instance of it yourself):
 
-`LIQUIDATOR_ADDRESS="0xABC..." yarn liquidation-bot`
+`LIQUIDATOR_ADDRESS="0xABC..." MARKET="01-USDC" yarn liquidation-bot --network kovan`
```
Comment (scott-silver) [inline README.md:231]: As far as I can tell, the only way to pass arguments to Hardhat scripts (anything run via `hardhat run ...`) is to use environment variables.

https://hardhat.org/hardhat-runner/docs/advanced/scripts#hardhat-arguments

It's definitely clunky to have to pass in some arguments via env variables and some (like `--network`) via params

File: hardhat.config.ts [hunk 3]
```diff
@@ -159,24 +159,31 @@ const config: HardhatUserConfig = {
...
+        "USDC-01": kovanUSDCRelationConfigMap,
+      }
     },
   },
 
   scenario: {
     bases: [
       {
         name: 'development',
+        market: '01-USDC' // ?
```
Comment (scott-silver) [inline hardhat.config.ts:175]: this is slightly awkward. I updated `ForkSpec` to require a `market` argument as well.

it seems like `market` should be required, but it means we need to add a market for the `development` base that will just be ignored.

File: plugins/deployment_manager/test/DeploymentManagerTest.ts [hunk 4]
```diff
@@ -93,23 +93,22 @@ describe('DeploymentManager', () => {
...
-        'avalanche'
       );
       expect(importResult).to.eql(fiatTokenBuildFile);
     });
   });
 
   describe('deploy', () => {
     it('should deploy succesfully', async () => {
-      let deploymentManager = new DeploymentManager('test', hre, {
+      let deploymentManager = new DeploymentManager('test', 'XXX', hre, {
```
Comment (scott-silver) [inline plugins/deployment_manager/test/DeploymentManagerTest.ts:111]: - [ ] update from `XXX` to `test-market`

File: plugins/deployment_manager/test/ImportTest.ts [hunk 5]
```diff
@@ -47,7 +47,7 @@ export const fiatTokenBuildFile: BuildFile = {
   version: 'v0.6.12+commit.27d51765',
 };
 
-export function mockImportSuccess(hre: HardhatRuntimeEnvironment, address: string) {
+export function mockImportSuccess(address: string) {
```
Comment (scott-silver) [inline plugins/deployment_manager/test/ImportTest.ts:50]: removed the `hre` argument from this function since it wasn't being used

File: MIGRATIONS.md [hunk 6]
```diff
@@ -8,10 +8,10 @@ Migrations are simple scripts which deploy or modify contracts. The goal of migr
 To create a new migration, run:
 
 ```sh
-yarn hardhat gen:migration --network kovan my_migration
+yarn hardhat gen:migration --network kovan --market 01-USDC my_migration
```
Comment (jflatow) [inline MIGRATIONS.md:11]: I know we discussed this already, but given that these may be part of URLs etc, I think we should drop the prefix and only add a suffix to differentiate if/when needed

File: deployments/fuji/01-USDC/deploy.ts [hunk 7]
```diff
@@ -1,15 +1,39 @@
+<<<<<<< HEAD:deployments/fuji/deploy.ts
```
Comment (scott-silver) [inline deployments/fuji/01-USDC/deploy.ts:1]: - [ ] delete rebase artifacts

File: plugins/deployment_manager/DeploymentManager.ts [hunk 8]
```diff
@@ -38,15 +35,17 @@ export class DeploymentManager {
...
     hre: HardhatRuntimeEnvironment,
     config: DeploymentManagerConfig = {}
   ) {
+    this.network = network;
     this.deployment = deployment;
     this.hre = hre;
     this.config = config;
 
-    this.cache = new Cache(deployment, config.writeCacheToDisk ?? false, config.baseDir);
+    this.cache = new Cache(`${this.network}/${this.deployment}`, config.writeCacheToDisk ?? false, config.baseDir);
```
Comment (scott-silver) [inline plugins/deployment_manager/DeploymentManager.ts:48]: this is slightly strange; previously you could derive the path to the roots.json file from the "deployment" alone.

the other option would be to update `Cache` to take in both network and deployment.

**Review by hayesgm — COMMENTED** [review]
Awesome, looks good to me.

File: plugins/deployment_manager/DeploymentManager.ts (hunk 8 above)
Comment (scott-silver) [reply plugins/deployment_manager/DeploymentManager.ts]: updated so that Cache takes a network and deployment

File: .github/workflows/run-spider.yaml (hunk 1 above)
Comment (jflatow) [reply .github/workflows/run-spider.yaml:1]: I think its gone already on main?

File: .github/workflows/prepare-migration.yaml [hunk 9]
```diff
@@ -42,13 +45,13 @@ jobs:
...
           ETH_PK: ${{ github.event.inputs.eth_pk }}
           DEBUG: true
 
       - uses: actions/upload-artifact@v2 # upload test results
         if: success() || failure() # run this step even if previous step failed
         with:
-          name: ${{ github.event.inputs.network }}-${{ github.event.inputs.migration }}
-          path: deployments/${{ github.event.inputs.network }}/artifacts/${{ github.event.inputs.migration }}.json
+          name: ${{ github.event.inputs.network }}-${{ github.event.inputs.deployment }}-${{ github.event.inputs.migration }}
+          path: deployments/${{ github.event.inputs.network }}/${{ github.event.inputs.deployment }}/artifacts/${{ github.event.inputs.migration }}.json
```
Comment (jflatow) [inline .github/workflows/prepare-migration.yaml:57]: Did you see the new `deploy-market` workflow?

File: hardhat.config.ts [hunk 10]
```diff
@@ -158,24 +158,31 @@ const config: HardhatUserConfig = {
...
 
   scenario: {
     bases: [
       {
         name: 'development',
+        deployment: 'usdc'
       },
       {
         name: 'kovan',
+        deployment: 'usdc',
```
Comment (jflatow) [inline hardhat.config.ts:178]: Can we get rid of the network params now that we have the network name for the deployment?

File: .github/workflows/prepare-migration.yaml (hunk 9 above)
Comment (scott-silver) [reply .github/workflows/prepare-migration.yaml:57]: missed that one. i'll update it

File: hardhat.config.ts (hunk 10 above)
Comment (scott-silver) [reply hardhat.config.ts:178]: we'll still need it for scenarios, right?

scenarios will iterate over `bases`; for each base it finds, it will generate an hre using the `name` field, and then look for a `roots.json` file in `${name}/${deployment}/roots.json`.

seems like we should rename `name` to `network`

Comment (jflatow) [reply hardhat.config.ts:178]: Well but the network is configured as part of the deployment now, that was the whole point of including it in the deployment info (encoded in path currently, alternatively was `network.json`)? So `base` has a name so that we can refer to it from `--bases`, and it uses a certain deployment, but should be able to pull chain args from the deployment network, right?

Comment (scott-silver) [reply hardhat.config.ts:178]: So if you've got two deployments:

`/deployments/kovan/usdc`
`/deployments/fuji/usdc`

and your bases are:

```
[
  {
    deployment: 'usdc'
  },
  {
    deployment: 'usdc'
  }
]
```

how would you specify that you only wanted to run the scenarios for `fuji/usdc`?

```
yarn scenario --bases usdc
```

or are you saying we should have the deployment name be `fuji/usdc`?

Comment (scott-silver) [reply hardhat.config.ts:178]: I guess it really needs to be `fuji/usdc` and `kovan/usdc`; right now, we're limited to running scenarios on a single deployment per network

File: .github/workflows/prepare-migration.yaml (hunk 9 above)
Comment (scott-silver) [reply .github/workflows/prepare-migration.yaml:57]: updated

File: hardhat.config.ts (hunk 10 above)
Comment (jflatow) [reply hardhat.config.ts:178]: It's the base name not the deployment name! Technically both deployments are named `usdc`, the bases are named whatever you name them in the hardhat config

File: hardhat.config.ts [hunk 11]
```diff
@@ -158,24 +158,31 @@ const config: HardhatUserConfig = {
...
+        usdc: kovanUSDCRelationConfigMap,
+      }
     },
   },
 
   scenario: {
     bases: [
       {
         name: 'development',
+        deployment: 'dai'
```
Comment (jflatow) [inline hardhat.config.ts:174]: Yay 🙌

**Review by jflatow — APPROVED** [review]
Great work 🥈 ! Can't wait to get this in, it's probably the last major reorganization we needed and should let the dependent repos (webb3, devops) build upon some solid ground 🏆

**Comment by scott-silver** [comment]
okay, this should be done now.

@jflatow this is the commit that updates our scenario bases, per our discussion yesterday:

https://github.com/compound-finance/comet/pull/498/commits/c85e8366ab07a101d71a172e802b82a68457c060

it involves renaming the `deployments/development/` to `deployments/hardhat` to match the convention of `deployment/${network}/${deployment}`

**Comment by jflatow** [comment]
> okay, this should be done now.
> 
> @jflatow this is the commit that updates our scenario bases, per our discussion yesterday:
> 
> [c85e836](https://github.com/compound-finance/comet/commit/c85e8366ab07a101d71a172e802b82a68457c060)
> 
> it involves renaming the `deployments/development/` to `deployments/hardhat` to match the convention of `deployment/${network}/${deployment}`

Ah nice but damn, was hoping we could keep it named `development`. Did you confirm scens are doing the right thing? Just thinking the network name might be `hardhat` for forked nodes to and since this didn't pop up before want to make sure its not doing something unexpected

**Comment by scott-silver** [comment]
Was able to restore `deployments/development`: https://github.com/compound-finance/comet/pull/498/commits/d015dcadabaabf8308c234252972da8ff8249288

It's determined by what we pass in to the `DeploymentManager` as the `network` argument (which then determines the path to write/read assets, e.g. `deployments/development/dai/roots.json`)

We've been special-casing the base with the name "development" in a few places. The one downside to continuing with this approach (as opposed to the original `deployments/hardhat/` approach) is that now we can only have a single development base.

Previously, we could have added an additional development base:

```
[
  {
    name: "development-base-1",
    network: "hardhat",
    deployment: "dai"
  },
  {
    name: "development-base-2",
    network: "hardhat",
    deployment: "usdc"
  }
]
````

So basically: if we're cool with there only being one `development` base, then we can just special-case based on the base name.

File: plugins/scenario/World.ts [hunk 12]
```diff
@@ -22,13 +22,14 @@ export class World {
...
-    this.deploymentManager = new DeploymentManager(base.name, hre, { debug: true });
+    const networkName = base.name === 'development' ? 'development' : base.network;
+    this.deploymentManager = new DeploymentManager(networkName, base.deployment, hre, { debug: true });
     this.snapshotDeploymentManager = this.deploymentManager;
   }
 
   // TODO: Can we do this better?
   isRemoteFork(): boolean {
-    return !!this.base.url;
+    return this.base.network !== 'hardhat';
```
Comment (jflatow) [inline plugins/scenario/World.ts:32]: If this is better (why) should we get rid of the TODO?

File: plugins/scenario/World.ts [hunk 13]
```diff
@@ -22,13 +22,14 @@ export class World {
     // Q: should we really need to fork/snapshot the deployment manager?
     this.hre = hre;
     this.base = base;
-    this.deploymentManager = new DeploymentManager(base.name, hre, { debug: true });
+    const networkName = base.name === 'development' ? 'development' : base.network;
```
Comment (jflatow) [inline plugins/scenario/World.ts:25]: `this.isRemoteFork() ? base.network : 'development';`?

File: plugins/scenario/World.ts (hunk 12 above)
Comment (scott-silver) [reply plugins/scenario/World.ts:31]: deleted

File: plugins/scenario/World.ts (hunk 13 above)
Comment (scott-silver) [reply plugins/scenario/World.ts]: that works. updated

File: plugins/scenario/World.ts [hunk 14]
```diff
@@ -22,12 +22,11 @@ export class World {
     // Q: should we really need to fork/snapshot the deployment manager?
     this.hre = hre;
     this.base = base;
-    const networkName = base.name === 'development' ? 'development' : base.network;
+    const networkName = this.isRemoteFork() ? base.network : 'development';
```
Comment (jflatow) [inline plugins/scenario/World.ts:25]: I thought we agreed in channel not to special case though?

Comment (scott-silver) [reply plugins/scenario/World.ts]: deleted

File: plugins/scenario/World.ts (hunk 13 above)
Comment (scott-silver) [reply plugins/scenario/World.ts]: deleted; we settled on using `deployments/hardhat`

**Review by jflatow — APPROVED** [review]
🔥
